Automattic/woocommerce-payments
# Fix race conditions leading to duplicate order status transitions and invoice already paid errors

Fixes #3273 

#### Changes proposed in this Pull Request

When processing a payment via the UPE there's a race condition whereby the `payment_intent.succeeded` webhook is received and handled by the site while the site is already in the process of handling the original payment request. 

Both requests get their own instance of the order object with a pending status and then attempt to complete the order (via `WC_Order->payment_complete()`).

You can find the two stack traces in my comment here: https://github.com/Automattic/woocommerce-payments/issues/3273#issuecomment-960413704

Because both requests complete the order (resulting in duplicate order notes), it also triggers two calls to the `woocommerce_order_payment_status_changed` action and the subsequent `WC_Payments_Invoice_Service::maybe_record_invoice_payment()` function. 

Calling `maybe_record_invoice_payment()` for the same order twice leads to us attempting to mark the subscription initial invoice as paid (out of band) twice and causes to the following 400 error.

```
"message": "Invoice is already paid",
"type": "invalid_request_error"
```

This PR fixes those issues.

**Questions/concerns**
In this PR I fix the race condition that was leading to the duplicate paid invoice, however as a fallback, if we attempt to pay an already paid invoice, instead of throwing the error and adding the order note, I catch that specific error on client side and log it. This means it's still theoretically possible that these "invoice is already paid" errors occur on server, although given they're 400 errors, they aren't logged to the Slack channel. What is your opinion on that? Should we do more to prevent those, or handle them better on the server? 

#### Testing Prereqs
- Listening to webhooks. `npm run listen` in your local server.
- UPE enabled.  
- WC Subscriptions plugin disabled.
- A plugin like Query Monitor to see HTTP request errors.

#### Testing instructions 
1. Purchase a subscription product using WC Pay and enter a new payment method on checkout.
2. On `develop` you'll notice the following errors:
   - On the order received page you should get a HTTP error (see below).
   - If you view the order in the admin dashboard you'll notice that there are 2 order notes suggesting the order transitioned from **Pending → Processing** twice. There's also a note about the invoice is already paid. See the screenshot below.
   - If you view the WC pay logs in **WooCommerce → Status → Logs** you'll also notice a `payment_intent not found in array in ...` error. I've included the full error below.
3. Repeating the purchase on this branch should resolve those issues. 

#### Screenshots
**HTTP API error**

<img width="2262" alt="Screen Shot 2021-11-04 at 4 05 27 pm" src="https://user-images.githubusercontent.com/8490476/140265873-c42e2faf-254f-4557-a6f3-ecd500fe3150.png">

**Order notes**

<img width="297" alt="Screen Shot 2021-11-04 at 4 09 58 pm" src="https://user-images.githubusercontent.com/8490476/140266323-564e0274-04b9-4d6f-8bb2-0301db652dbf.png">


**`woocommerce-payments` log error**
```
ERROR WCPay\Exceptions\Rest_Request_Exception: payment_intent not found in array in wp-content/plugins/woocommerce-payments/includes/subscriptions/class-wc-payments-subscriptions-event-handler.php:197
Stack trace:
#0 wp-content/plugins/woocommerce-payments/includes/subscriptions/class-wc-payments-subscriptions-event-handler.php(97): WC_Payments_Subscriptions_Event_Handler->get_event_property(Array, 'payment_intent')
#1 wp-content/plugins/woocommerce-payments/includes/admin/class-wc-rest-payments-webhook-controller.php(137): WC_Payments_Subscriptions_Event_Handler->handle_invoice_paid(Array)
#2 wp-includes/rest-api/class-wp-rest-server.php(1140): WC_REST_Payments_Webhook_Controller->handle_webhook(Object(WP_REST_Request))
#3 wp-includes/rest-api/class-wp-rest-server.php(987): WP_REST_Server->respond_to_request(Object(WP_REST_Request), '/wc/v3/payments...', Array, NULL)
#4 wp-includes/rest-api/class-wp-rest-server.php(414): WP_REST_Server->dispatch(Object(WP_REST_Request))
#5 wp-includes/rest-api.php(370): WP_REST_Server->serve_request('/wc/v3/payments...')
#6 wp-includes/class-wp-hook.php(303): rest_api_loaded(Object(WP))
#7 wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters(NULL, Array)
#8 wp-includes/plugin.php(518): WP_Hook->do_action(Array)
#9 wp-includes/class-wp.php(388): do_action_ref_array('parse_request', Array)
#10 wp-includes/class-wp.php(750): WP->parse_request('')
#11 wp-includes/functions.php(1291): WP->main('')
#12 wp-blog-header.php(16): wp()
#13 index.php(17): require('/Users/jamesall...')
#14 {main}
```

-------------------


**Post merge**
